docs: add api reference section (#716)
software-mansion/react-native-enriched: docs/docs/api-reference/enriched-text-input.md

# EnrichedTextInput

`EnrichedTextInput` is a rich text editor that styles text live as you type.
It is uncontrolled — content lives on the native side and you talk to it through

> a `ref`. See [Core concepts](/fundamentals/core-concepts) for the mental model.

Reviewer: NIT: Maybe let's shorten this? I feel like api-reference section should be more straight-forward and that description can already be found in the `core-concepts`
```suggestion
`EnrichedTextInput` is a rich text editor that styles text live as you type.
See [Core concepts](/fundamentals/core-concepts) for the mental model.
```

  linkRegex?: RegExp | null;
  returnKeyType?: ReturnKeyTypeOptions;
  returnKeyLabel?: string;
  submitBehavior?: 'submit' | 'blurAndSubmit' | 'newline';
  onFocus?: (e: FocusEvent) => void;
  onBlur?: (e: BlurEvent) => void;
  onChangeText?: (e: NativeSyntheticEvent<OnChangeTextEvent>) => void;
  onChangeHtml?: (e: NativeSyntheticEvent<OnChangeHtmlEvent>) => void;
  onChangeState?: (e: NativeSyntheticEvent<OnChangeStateEvent>) => void;
  onLinkDetected?: (e: OnLinkDetected) => void;
  onMentionDetected?: (e: OnMentionDetected) => void;
  onStartMention?: (indicator: string) => void;
  onChangeMention?: (e: OnChangeMentionEvent) => void;
  onEndMention?: (indicator: string) => void;
  onChangeSelection?: (e: NativeSyntheticEvent<OnChangeSelectionEvent>) => void;
  onKeyPress?: (e: NativeSyntheticEvent<OnKeyPressEvent>) => void;
  onSubmitEditing?: (e: NativeSyntheticEvent<OnSubmitEditing>) => void;
  onPasteImages?: (e: NativeSyntheticEvent<OnPasteImagesEvent>) => void;
  contextMenuItems?: ContextMenuItem[];
  textShortcuts?: TextShortcut[];
  androidExperimentalSynchronousEvents?: boolean;
  useHtmlNormalizer?: boolean;
  allowFontScaling?: boolean;
}
```

</details>

## Props

> ### `allowFontScaling` <Optional /> {#allowfontscaling}

Reviewer: I'm not sure about that `Optional` mark. Pretty much every prop we support is optional and IMO it creates unnecessary noise when reading, especially as its style it is pretty vibrant. What about changing the approach and marking the non-optional fields with a `<Required />` instead?

Author: Lets wait and see what @szydlovsky and @exploIF think about it ;)

Reviewer 2: @hejsztynx @kacperzolkiewski I agree that most if not of all the props we have are optional. Do I see it right that the `EnrichedText`'s `children` is the only required prop? If so, I'm okay with using just a single `<Required />` and also emphasizing this fact in the docs.

Author: Yes, you are right. I removed all the `Optional` badges and added a `Required` badge for `children` in `EnrichedText`.

- `sentences` — first letter of each sentence
- `none` — don't auto-capitalize anything

| Type                                               | Default       | Platforms         |
| -------------------------------------------------- | ------------- | ----------------- |
| `'none' \| 'sentences' \| 'words' \| 'characters'` | `'sentences'` | Android, iOS, Web |

### `contextMenuItems` <Optional /> {#contextmenuitems}

An array of custom items to display in the native text editing menu. Each item
specifies a title, visibility flag, and a callback that fires when the item is
tapped.

The `onPress` callback receives a single object argument with:

- `text` — the currently selected text
- `selection` — an object with `start` and `end` indices of the current selection
- `styleState` — the latest `OnChangeStateEvent` payload reflecting active styles
  at the time of the tap

```ts
interface ContextMenuItem {
  text: string;
  onPress: (args: {
    text: string;
    selection: { start: number; end: number };
    styleState: OnChangeStateEvent;
  }) => void;
  visible?: boolean;
}
```

- `text` is the title displayed in the menu
- `onPress` is the callback invoked when the item is tapped

> - `visible` controls whether the item is shown; defaults to `true`

Reviewer: Maybe let's move the details of `onPress` after it is introduced
```suggestion
```ts
interface ContextMenuItem {
  text: string;
  onPress: (args: {
    text: string;
    selection: { start: number; end: number };
    styleState: OnChangeStateEvent;
  }) => void;
  visible?: boolean;
}
```

- `text` is the title displayed in the menu
- `onPress` is the callback invoked when the item is tapped
- `visible` controls whether the item is shown; defaults to `true`

The `onPress` callback receives a single object argument with:

- `text` — the currently selected text
- `selection` — an object with `start` and `end` indices of the current selection
- `styleState` — the latest `OnChangeStateEvent` payload reflecting active styles
  at the time of the tap
```

via ref methods still work.

:::

### `htmlStyle` <Optional /> {#htmlstyle}

Customizes the appearance of HTML elements inside the editor. See
[`HtmlStyle`](#htmlstyle-type).

| Type        | Default                                            | Platforms         |
| ----------- | -------------------------------------------------- | ----------------- |
| `HtmlStyle` | default values from [`HtmlStyle`](#htmlstyle-type) | Android, iOS, Web |

### `mentionIndicators` <Optional /> {#mentionindicators}

The recognized mention indicators. Each item must be a 1-character string.

| Type       | Default | Platforms         |
| ---------- | ------- | ----------------- |
| `string[]` | `['@']` | Android, iOS, Web |

### `linkRegex` <Optional /> {#linkregex}

A custom regex pattern for detecting links in the input. If not provided, a
default regex is used. You can customize which patterns are recognized as links
— for example only `https://` URLs, or custom schemes.

Not all JS regex features are supported; for example variable-width lookbehinds
won't work.

| Type             | Default                       | Platforms    |
| ---------------- | ----------------------------- | ------------ |

> | `RegExp \| null` | default native platform regex | Android, iOS |

Reviewer: I believe it is implemented on web

```suggestion
| `RegExp \| null` | default native platform regex | Android, iOS, Web |
```

:::tip

Pass `null` to disable link detection completely.

:::

### `onBlur` <Optional /> {#onblur}

Called whenever the input loses focus.

| Type                     | Default | Platforms         |
| ------------------------ | ------- | ----------------- |
| `(e: BlurEvent) => void` | —       | Android, iOS, Web |

### `onChangeHtml` <Optional /> {#onchangehtml}

Called when the input's HTML changes.

```ts
interface OnChangeHtmlEvent {
  value: string;
}
```

- `value` is the new HTML

| Type                                                       | Default | Platforms         |
| ---------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnChangeHtmlEvent>) => void` | —       | Android, iOS, Web |

:::tip

Specifying `onChangeHtml` may have performance implications, especially with
large documents, as it requires continuous HTML parsing. If you only need the
HTML at specific moments (for example when saving), use the
[`getHTML`](#gethtml) ref method instead. When `onChangeHtml` is not provided,

> the component avoids unnecessary HTML parsing.

Reviewer: It feels like it's already written in the first sentence

```suggestion
HTML at specific moments (for example when saving), use the
[`getHTML`](#gethtml) ref method instead.
```

:::

### `onChangeMention` <Optional /> {#onchangemention}

> Called whenever the user changes a mention that is being edited.

Reviewer: Wording feels wrong, as it could imply the change of edited mentions, as switching between two different ones. Where `onChangeMention` fires when the query/text inside a single mention changes.

| ---------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnChangeTextEvent>) => void` | —       | Android, iOS, Web |

:::tip

If you don't need the plain text value, omit `onChangeText` — continuous text
extraction can have performance implications.

:::

### `onEndMention` <Optional /> {#onendmention}

Called when the user is no longer editing a mention actively — they moved the
cursor elsewhere, or typed a space and the cursor is no longer within the edited
mention.

- `indicator` is the indicator of the mention that was being edited

| Type                          | Default | Platforms         |
| ----------------------------- | ------- | ----------------- |
| `(indicator: string) => void` | —       | Android, iOS, Web |

### `onFocus` <Optional /> {#onfocus}

Called whenever the input is focused.

| Type                      | Default | Platforms         |
| ------------------------- | ------- | ----------------- |
| `(e: FocusEvent) => void` | —       | Android, iOS, Web |

### `onLinkDetected` <Optional /> {#onlinkdetected}

Called when a new link has been added, or the user has moved the

> cursor/selection onto a link.

Reviewer: that's not 100% true. I believe it doesn't fire on a pure link addition, but rather from the fact that after it is added, the cursor sits in it. E.g. we paste some text which has a link, then no `onLinkDetected` is fired.

```suggestion
Called when the user has moved the
cursor/selection onto a link.
```

```ts
interface OnLinkDetected {
  text: string;
  url: string;
  start: number;
  end: number;
}
```

- `text` is the link's displayed text
- `url` is the underlying URL
- `start` is the starting index of the link
- `end` is the first index after the ending index of the link

| Type                              | Default | Platforms         |
| --------------------------------- | ------- | ----------------- |
| `(event: OnLinkDetected) => void` | —       | Android, iOS, Web |

### `onMentionDetected` <Optional /> {#onmentiondetected}

Called when a mention has been detected — either a new mention was added, or the

> user moved the cursor/selection onto a mention.

Reviewer: The same case with `onLinkDetected`
```suggestion
Called when the
user moved the cursor/selection onto a mention.
```

```ts
interface OnMentionDetected {
  text: string;
  indicator: string;
  attributes: Record<string, string>;
}
```

- `text` is the mention's displayed text
- `indicator` is the indicator of the mention
- `attributes` are the additional user-defined attributes stored with the mention

| Type                                 | Default | Platforms         |
| ------------------------------------ | ------- | ----------------- |
| `(event: OnMentionDetected) => void` | —       | Android, iOS, Web |

### `onStartMention` <Optional /> {#onstartmention}

> Called whenever mention editing starts (after placing the indicator).

Reviewer: It also fires when we re-enter an unfinished mention.

```suggestion
Called whenever mention editing starts.
```

  key: string;
}
```

| Type                                                     | Default | Platforms         |
| -------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnKeyPressEvent>) => void` | —       | Android, iOS, Web |

### `onSubmitEditing` <Optional /> {#onsubmitediting}

Called when the user submits the input (presses the return/enter key while
`submitBehavior` is `'submit'` or `'blurAndSubmit'`).

```ts
interface OnSubmitEditing {
  text: string;
}
```

- `text` is the current plain-text content of the input at submission time

| Type                                                     | Default | Platforms         |
| -------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnSubmitEditing>) => void` | —       | Android, iOS, Web |

### `onPasteImages` <Optional /> {#onpasteimages}

Called when the user pastes one or more images or GIFs into the input.

- `images` — an array of objects with URI, MIME type, and dimensions for each
  pasted image/GIF
- **Web:** each `uri` is a `blob:` URL (`URL.createObjectURL`). If you retain

>   URIs, call `URL.revokeObjectURL` when finished so blobs can be released

Reviewer: that's already in the `:::note` below

and maybe we should use `:::caution` instead? But I'm not sure about that though
```suggestion
```

  | 'ordered_list'
  | 'checkbox_list';
```

- `trigger` is the typed pattern that activates the shortcut
- `style` is the style to apply when the trigger completes

**[Paragraph styles](/fundamentals/html-format-and-supported-tags#paragraph-tags)**
fire at the start of a paragraph (e.g. `# ` → H1, `- ` → unordered list).
Supported styles: `h1`–`h6`, `blockquote`, `codeblock`, `unordered_list`,
`ordered_list`, `checkbox_list`.

:::note

Paragraph shortcuts are only effective on plain paragraphs. If the paragraph
already has an active paragraph style (for example it is already a heading or a
list item), typing the trigger pattern has no effect.

:::

**[Inline styles](/fundamentals/html-format-and-supported-tags#inline-tags)**
fire when a closing delimiter is typed around text (e.g. `**text**` → bold). The
trigger is the delimiter string (e.g. `**`, `*`, `~~`). Supported styles:
`bold`, `italic`, `underline`, `strikethrough`, `inline_code`.

:::note

Style rules still apply to shortcut-triggered styles: if the target style is
**blocked** by another currently active style (e.g. bold inside a codeblock),
the shortcut has no effect. If the target style **conflicts** with another

> active style, the conflicting style is removed when the new one is applied. See

Reviewer: Currently this might imply that e.g. `h1` shortcut in a `h2` paragraph, will conflict, and replace the `h2` into `h1`. Maybe it's just worth clarifying that the conflicting behavior only applies to inline style shortcuts. But honestly I'm not sure 🤔 

Also I wrote similar stuff in the #713 `Text shortcuts` section, I wonder where that should ultimately be present. What do you think?

Author: OK but few lines above this section we have:
```
:::note

Paragraph shortcuts are only effective on plain paragraphs. If the paragraph
already has an active paragraph style (for example it is already a heading or a
list item), typing the trigger pattern has no effect.

:::
```

Reviewer: I know, but it might be worth clarifying, as I got confused by it myself, when I was reading that. If you see it as sufficient, I see your point, I don't insist ;)

Author: Fixed

```ts
setImage: (src: string, width: number, height: number) => void;
```

Sets an inline image at the current selection.

- `src: string` — absolute path to a file or remote image address
- `width: number` — width of the image
- `height: number` — height of the image

:::note

It's the developer's responsibility to provide proper width and height, which
may require calculating aspect ratio. If the image source is incorrect, a static
placeholder is displayed.

:::

### `.setLink()`

```ts
setLink: (start: number, end: number, text: string, url: string) => void;
```

Sets a link at the given place with the given displayed text and URL. The link
replaces any text between `start` and `end`. Setting a link with `start` equal
to `end` inserts it in place.

- `start: number` — starting index where the link should be
- `end: number` — first index behind the new link's ending index
- `text: string` — displayed text of the link
- `url: string` — URL of the link

### `.removeLink()`

```ts
removeLink: (start: number, end: number) => void;
```

Removes link styling from any links found within the given range. The text
content is preserved; only the link attributes are stripped. Out-of-bounds

> values are clamped to a valid range.

Reviewer: NIT: I wonder if that's relevant and worth mentioning.

```suggestion
content is preserved; only the link attributes are stripped.
```

- `start: number` — starting index of the range to remove links from
- `end: number` — first index behind the range's ending index

### `.setMention()`

```ts
setMention: (
  indicator: string,
  text: string,
  attributes?: Record<string, string>
) => void;
```

Sets the currently edited mention with a given indicator, displayed text, and
custom attributes.

- `indicator: string` — indicator of the set mention
- `text: string` — text displayed for the mention; anything the user typed is
  replaced by that text. The mention indicator isn't added to that text
- `attributes?: Record<string, string>` — additional custom attributes for the
  mention, preserved through parsing to and from HTML

> 

Reviewer: It might be nice to add a note about using `data-` preifx

for example (I added such note to mantion-only-input guide):

```suggestion

:::note

The attributes you pass to `setMention` ride along in the HTML
and survive a round-trip through `getHTML` / `setValue`. Prefix custom keys with
`data-` if they need to outlive a sanitizer - see the note in
[Mentions](/rich-text-formatting/mentions).

:::

```

  defaults to `darkgray`
- `borderWidth` — width of that border; defaults to `4`
- `gapWidth` — width of the gap between the border and the blockquote text;
  defaults to `16`
- `color` — color of blockquote text; takes a
  [color](https://reactnative.dev/docs/colors) value; if not set, uses the
  input's [color](#style)

### codeblock

- `color` — color of codeblock text; takes a
  [color](https://reactnative.dev/docs/colors) value, defaults to `black`
- `borderRadius` — radius of the codeblock's border; defaults to `8`
- `backgroundColor` — codeblock background color; takes a
  [color](https://reactnative.dev/docs/colors) value, defaults to `darkgray`

### code (inline code)

- `color` — color of inline code text; takes a
  [color](https://reactnative.dev/docs/colors) value, defaults to `red`
- `backgroundColor` — inline code background color; takes a
  [color](https://reactnative.dev/docs/colors) value, defaults to `darkgray`

### a (link)

- `color` — color of link text; takes a
  [color](https://reactnative.dev/docs/colors) value, defaults to `blue`
- `textDecorationLine` — whether links are underlined; `'underline'` or
  `'none'`, defaults to `'underline'`

### mention

> 

Reviewer: we should probably mention the `'default'` style, that user can create. I added this in `styling-the-input` section:

```
:::tip

You can also create a default `mention` style config, by using the `'default'` key.

htmlStyle={{
  mention: {
    'default': { color: '#2563eb', backgroundColor: '#dbeafe' },
    '#': { color: '#16a34a', backgroundColor: '#dcfce7' },
  },
}}

This way you can create a style for any mention indicator to fallback if it doesn't have one fully defined.

:::
```

the same would apply to the `EnrichedText` api ref

  filter?: TextStyle['filter'];
  /** @platform android web */
  mixBlendMode?: TextStyle['mixBlendMode'];
  opacity?: TextStyle['opacity'];
  /** @platform ios web */
  outlineColor?: ColorValue;
  outlineOffset?: TextStyle['outlineOffset'];
  /** @platform android web */
  outlineStyle?: TextStyle['outlineStyle'];
  outlineWidth?: TextStyle['outlineWidth'];
  /** @platform ios web */
  pointerEvents?: TextStyle['pointerEvents'];

  // Typography
  color?: ColorValue;
  fontFamily?: string;
  fontSize?: number;
  fontStyle?: TextStyle['fontStyle'];
  fontWeight?: TextStyle['fontWeight'];
  lineHeight?: number;
  /** @platform web */
  letterSpacing?: number;
}
```

</details>

## Remarks

- The input is uncontrolled. Change content and formatting through ref methods;
  observe changes through events. See
  [Core concepts](/fundamentals/core-concepts#the-input-is-uncontrolled).
- Prefer [`getHTML`](#gethtml) over continuous `onChangeHtml` when you only need
  HTML at specific moments.
- Sanitizing HTML is your responsibility. See
  [Core concepts](/fundamentals/core-concepts#html-is-the-source-of-truth).
- For the full list of supported tags and style conflicts, see

>   [Supported tags](/fundamentals/html-format-and-supported-tags).

Reviewer: Maybe this would be better?
```suggestion
  [HTML format and supported tags](/fundamentals/html-format-and-supported-tags).
```

- `text` is the current plain-text content of the input at submission time

| Type                                                     | Default | Platforms         |
| -------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnSubmitEditing>) => void` | -       | Android, iOS, Web |

### `onPasteImages` <Optional /> {#onpasteimages}

Called when the user pastes one or more images or GIFs into the input.

- `images` - an array of objects with URI, MIME type, and dimensions for each
  pasted image/GIF

```ts
interface OnPasteImagesEvent {
  images: {
    uri: string;
    type: string;
    width: number;
    height: number;
  }[];
}
```

| Type                                                        | Default | Platforms         |
| ----------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnPasteImagesEvent>) => void` | -       | Android, iOS, Web |

:::note

On Web, `uri` is a blob URL (`blob:...`). Blob URLs hold memory until explicitly
released. Call `URL.revokeObjectURL(uri)` once you no longer need the image
(for example after the upload completes).

> :::

Reviewer 2: I'd make it a tip, as it rather suggests some good programming practice rather than explains some unusual or confusing API implications.

list item), typing the trigger pattern has no effect.

:::

**[Inline styles](/fundamentals/html-format-and-supported-tags#inline-tags)**
fire when a closing delimiter is typed around text (e.g. `**text**` → bold). The
trigger is the delimiter string (e.g. `**`, `*`, `~~`). Supported styles:
`bold`, `italic`, `underline`, `strikethrough`, `inline_code`.

:::note

Style rules still apply to shortcut-triggered styles: if the target style is
**blocked** by another currently active style (e.g. bold inside a codeblock),
the shortcut has no effect. If the target style **conflicts** with another
active inline style, the conflicting style is removed when the new one is
applied. See [Supported tags](/fundamentals/html-format-and-supported-tags) for the full
conflict and blocking rules.

:::

Default value:

```ts
[
  { trigger: '- ', style: 'unordered_list' },
  { trigger: '1. ', style: 'ordered_list' },
];
```

| Type             | Default   | Platforms         |
| ---------------- | --------- | ----------------- |
| `TextShortcut[]` | see above | Android, iOS, Web |

:::note

Pass an empty array to disable all shortcuts.

> :::

Reviewer 2: Also should be a `tip` IMO.

### `.removeLink()`

```ts
removeLink: (start: number, end: number) => void;
```

Removes link styling from any links found within the given range. The text
content is preserved; only the link attributes are stripped.

- `start: number` - starting index of the range to remove links from
- `end: number` - first index behind the range's ending index

### `.setMention()`

```ts
setMention: (
  indicator: string,
  text: string,
  attributes?: Record<string, string>
) => void;
```

Sets the currently edited mention with a given indicator, displayed text, and
custom attributes.

- `indicator: string` - indicator of the set mention
- `text: string` - text displayed for the mention; anything the user typed is
  replaced by that text. The mention indicator isn't added to that text
- `attributes?: Record<string, string>` - additional custom attributes for the
  mention, preserved through parsing to and from HTML

:::note

The attributes you pass to `setMention` ride along in the HTML
and survive a round-trip through `getHTML` / `setValue`. Prefix custom keys with
`data-` if they need to outlive a sanitizer - see the note in
[Mentions](/rich-text-formatting/mentions).

> :::

Reviewer 2: Should be a `tip` or even a `caution` IMO - user that doesn't know this will suddenly lose their custom mention attributes.

| -------------------------------------------------- | ------------- | ----------------- |
| `'none' \| 'sentences' \| 'words' \| 'characters'` | `'sentences'` | Android, iOS, Web |

### `contextMenuItems` {#contextmenuitems}

An array of custom items to display in the native text editing menu. Each item
specifies a title, visibility flag, and a callback that fires when the item is
tapped.

```ts
interface ContextMenuItem {
  text: string;
  onPress: (args: {
    text: string;
    selection: { start: number; end: number };
    styleState: OnChangeStateEvent;
  }) => void;
  visible?: boolean;
}
```

- `text` is the title displayed in the menu
- `onPress` is the callback invoked when the item is tapped
- `visible` controls whether the item is shown; defaults to `true`

The `onPress` callback receives a single object argument with:

- `text` - the currently selected text
- `selection` - an object with `start` and `end` indices of the current selection
- `styleState` - the latest `OnChangeStateEvent` payload reflecting active styles
  at the time of the tap

| Type                | Default | Platforms    |
| ------------------- | ------- | ------------ |

> | `ContextMenuItem[]` | -       | Android, iOS |

Reviewer 3: That's a general note, but let's remove `Default` column for callback, there is no option to have a default callback implemented

Author: Removed from all callbacks

| --------------------------------------- | ------- | ----------------- |
| `(event: OnChangeMentionEvent) => void` | -       | Android, iOS, Web |

### `onChangeSelection` {#onchangeselection}

Called each time the user changes the selection or moves the cursor.

```ts
interface OnChangeSelectionEvent {
  start: number;
  end: number;
  text: string;
}
```

- `start` is the index of the selection's beginning
- `end` is the first index after the selection's ending; for a cursor with no
  selection, `start` equals `end`
- `text` is the input's text in the current selection

| Type                                                            | Default | Platforms         |
| --------------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnChangeSelectionEvent>) => void` | -       | Android, iOS, Web |

### `onChangeState` {#onchangestate}

Called when any of the styles within the selection changes. Use this to drive
toolbar button state. See
[The style state model](/fundamentals/core-concepts#the-style-state-model).

```ts
interface OnChangeStateEvent {

>   bold: { isActive: boolean; isConflicting: boolean; isBlocking: boolean };

Reviewer 3: Can we make it more readable? I think it makes sense to extract `{ isActive: boolean; isConflicting: boolean; isBlocking: boolean }` to a separate interface and reuse it whenever possible. Currently it's hard to see difference between styles

Author: I extracted `{ isActive: boolean; isConflicting: boolean; isBlocking: boolean }` into `StyleState` interface

  `'left'`, `'center'`, `'right'`, `'justify'`, or `'auto'`

:::note

On Android, `'justify'` is not supported. It is accepted in the type signature
but has no justified layout effect - text is shown with natural alignment
instead, the same as `'auto'`.

:::

| Type                                                        | Default | Platforms         |
| ----------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnChangeStateEvent>) => void` | -       | Android, iOS, Web |

### `onChangeText` {#onchangetext}

Called when any text changes occur in the input.

```ts
interface OnChangeTextEvent {
  value: string;
}
```

- `value` is the new plain-text value of the input

| Type                                                       | Default | Platforms         |
| ---------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnChangeTextEvent>) => void` | -       | Android, iOS, Web |

:::tip

If you don't need the plain text value, omit `onChangeText` - continuous text
extraction can have performance implications.

> :::

Reviewer 3: Let's remove it. It's true for html, but impact when using raw text should be minimal, not worth mentioning it imo

Author: done

Called when the user submits the input (presses the return/enter key while
`submitBehavior` is `'submit'` or `'blurAndSubmit'`).

```ts
interface OnSubmitEditing {
  text: string;
}
```

- `text` is the current plain-text content of the input at submission time

| Type                                                     | Default | Platforms         |
| -------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnSubmitEditing>) => void` | -       | Android, iOS, Web |

### `onPasteImages` {#onpasteimages}

Called when the user pastes one or more images or GIFs into the input.

- `images` - an array of objects with URI, MIME type, and dimensions for each
  pasted image/GIF

```ts
interface OnPasteImagesEvent {
  images: {
    uri: string;
    type: string;
    width: number;
    height: number;
  }[];
}
```

| Type                                                        | Default | Platforms         |
| ----------------------------------------------------------- | ------- | ----------------- |
| `(event: NativeSyntheticEvent<OnPasteImagesEvent>) => void` | -       | Android, iOS, Web |

> :::tip

Reviewer 3: Let's make it `note` instead

Author: done

  opacity?: TextStyle['opacity'];
  /** @platform ios web */
  outlineColor?: ColorValue;
  outlineOffset?: TextStyle['outlineOffset'];
  /** @platform android web */
  outlineStyle?: TextStyle['outlineStyle'];
  outlineWidth?: TextStyle['outlineWidth'];
  /** @platform ios web */
  pointerEvents?: TextStyle['pointerEvents'];

  // Typography
  color?: ColorValue;
  fontFamily?: string;
  fontSize?: number;
  fontStyle?: TextStyle['fontStyle'];
  fontWeight?: TextStyle['fontWeight'];
  lineHeight?: number;
  /** @platform web */
  letterSpacing?: number;
}
```

</details>

## Remarks

- The input is uncontrolled. Change content and formatting through ref methods;
  observe changes through events. See
  [Core concepts](/fundamentals/core-concepts#the-input-is-uncontrolled).
- Prefer [`getHTML`](#gethtml) over continuous `onChangeHtml` when you only need
  HTML at specific moments.
- Sanitizing HTML is your responsibility. See
  [Core concepts](/fundamentals/core-concepts#html-is-the-source-of-truth).
- For the full list of supported tags and style conflicts, see
  [HTML format and supported tags](/fundamentals/html-format-and-supported-tags).

> ## Platform compatibility

Reviewer 3: I think we can remove it
